src/v/storage/multi_snapshot.h

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "bytes/iobuf.h"
+#include "model/fundamental.h"
+#include "seastarx.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/util/log.hh>
+
+#include <filesystem>
+
+namespace storage {
+
+class snapshot_reader;
+class snapshot_writer;
+
+/**
+ * Snapshot manager.
+ *
+ * One instance of the manager handles the life cycle of snapshots for a single
+ * raft group. Snapshots are organized into a directory:
+ *
+ *    <dir>/
+ *      - snapshot              <- current snapshot (optional)
+ *      - snapshot.partial.ts   <- partially written snapshot
+ *      - ...
+ *
+ * Snapshots are initially created as `snapshot.partial.*` files. When a
+ * snapshot is fully created and flushed to disk it is atomically renamed to
+ * become the current snapshot. Partial snapshots may accumulate when crashes
+ * occur, and the manager can be used to clean-up old snapshots.
+ *
+ * Snapshot file format is an envelope for the state machine snapshot data:
+ *
+ *    [ {version, crc}, metadata, <snapshot blob>, crc]
+ *
+ *    NOTE: the footer crc is not yet implemented.
+ *
+ * The version and crc are contained in snapshot_header, and are followed by
+ * snapshot metadata that can be used to store information about the snapshot
+ * and is accessible independent of the full snapshot data.  The crc covers
+ * everything from the beginning of the snapshot file up to the snapshot blob.
+ * This is all managed by the consensus and the state machine middleare.  The
+ * blob is written by the state machine itself. For example, a state machine may
+ * write a small serialized blob, or stream data to the snapshot file. Streaming
+ * may be useful in cases where the data structure permits determistic iteration
+ * such as a pre-order tree travseral. The snapshot file contains a footer that
+ * records a crc of the snapshot blob itself.
+ *
+ * Usage:
+ *
+ *    Create a manager instance:
+ *
+ *       multi_snapshot_manager mgr(prefix, "/path/to/ntp/", io_priority);
```

Comment:
It's needed to avoid collision when several snapshot_managers points to the same directory. In this case when one manager does `remove_partial_snapshots` it may accidentally remove a wip snapshot of another manager.

Reply:
I would have thought that if users needed to choose non conflicting filenames that that uniqueness would serve to disambiguate things.

Reply:
It isn't about uniqueness. snapshot_manager works by writing data to a temporary file and then moving it to a target snapshot. When an app crushes before moving it may lead to keeping the temporary file (we clean them with remove_partial_snapshots). Since managers don't have memory they can't distinguish its former snapshots from before the crush from the snapshots of concurrent managers. So we need something to help them to identify their temporary files.